internal/api/middleware/identity.go

```diff
@@ -1,26 +1,181 @@
 package middleware
 
 import (
+	"context"
 	"encoding/base64"
-	"encoding/json"
+	"fmt"
 	"net/http"
+	"os"
 
 	"github.com/labstack/echo/v4"
 	"github.com/redhatinsights/platform-go-middlewares/identity"
+	authv1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
 )
 
-func Identity(next echo.HandlerFunc) echo.HandlerFunc {
+const (
+	// ID Provider config values
+	rhSSOIDProvider = "rhsso"
+	oAuthIDProvider = "oauth"
+
+	// ID Provider header values
+	RHSSOIdentityHeader = "X-Rh-Identity"
+	OAuthIdentityHeader = "Bearer Token"
+)
+
+type IdentityProvider interface {
+	GetHandlerFunction() echo.MiddlewareFunc
+}
+
+type OAuthIdentityProvider struct {
+	client kubernetes.Interface
+}
+
+func NewOauthIDProvider(kubeClient kubernetes.Interface) IdentityProvider {
+	return &OAuthIdentityProvider{client: kubeClient}
+}
+
+func (o *OAuthIdentityProvider) GetHandlerFunction() echo.MiddlewareFunc {
+	return o.oauthIdentityHandlerFunction
+}
+
+func (o *OAuthIdentityProvider) oauthIdentityHandlerFunction(next echo.HandlerFunc) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		decodedIdentity, err := decodeIdentity(c, OAuthIdentityHeader)
+		if err != nil {
+			return err
+		}
+		token, err := identity.NewOAuthFromHeader(decodedIdentity)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusUnauthorized, fmt.Sprintf("Unable to marshal %s into struct", OAuthIdentityHeader))
+		}
+		if token == "" {
+			return echo.NewHTTPError(http.StatusUnauthorized, fmt.Sprintf("Missing %s", OAuthIdentityHeader))
+		}
+
+		userInfo, err := o.validateToken(string(token))
+		if err != nil {
+			return err
+		}
+		if userInfo == nil {
+			return echo.NewHTTPError(http.StatusUnauthorized, "User information is missing from TokenReview API")
+		}
+
+		c.Set("Identity", *userInfo)
+		return next(c)
+	}
+}
+
+func (o *OAuthIdentityProvider) validateToken(token string) (*authv1.UserInfo, error) {
+	// Validate the bearer token using Kubernetes TokenReview API
+	tokenReview := &authv1.TokenReview{
+		Spec: authv1.TokenReviewSpec{
+			Token: token,
+		},
+	}
+
+	result, err := o.client.AuthenticationV1().TokenReviews().Create(
```

Comment:
How should the token be defined? Perhaps you can add a step for describing it in some readme (even at org wiki level) so we'll know how to run this entire thing.

Reply:
What do you mean by `token be defined`? The previous lines create the token wrapper:

```
tokenReview := &authv1.TokenReview{
		Spec: authv1.TokenReviewSpec{
			Token: token,
		},
	}
```

And the token is the value from the `Bearer Token` header. Is this what you mean? to include a brief description where the token comes from?

Reply:
Pretty much that. the Bearer Token for instance, needs to be associated with Role to permit this token creation IINW.